src/main/java/org/killbill/billing/plugin/dwolla/api/DwollaPaymentPluginApi.java

```diff
@@ -0,0 +1,273 @@
+                    TransactionType.PURCHASE,
+                    amount,
+                    currency.toString(),
+                    errorMsg, // gatewayError
+                    errorCode, // gatewayErrorCode
+                    Lists.newArrayList(properties) // TODO PluginProperties.buildPluginProperties(...)
+            );
+        } catch (final SQLException e) {
+            logService.log(LogService.LOG_ERROR, e.getMessage());
+            throw new PaymentPluginApiException("Payment went through, but we encountered a database error. Payment details: " + (transfer.toString()), e);
+        }
+    }
+
+    @Override
+    public PaymentTransactionInfoPlugin voidPayment(UUID kbAccountId, UUID kbPaymentId, UUID kbTransactionId, UUID kbPaymentMethodId, Iterable<PluginProperty> properties, CallContext context) throws PaymentPluginApiException {
+        // TODO Does Dwolla have cancel operation?
+        return new DwollaPaymentTransactionInfoPlugin(kbPaymentId, kbTransactionId, TransactionType.VOID, null, null, Lists.newArrayList(properties));
+    }
+
+    @Override
+    public PaymentTransactionInfoPlugin creditPayment(UUID kbAccountId, UUID kbPaymentId, UUID kbTransactionId, UUID kbPaymentMethodId, BigDecimal amount, Currency currency, Iterable<PluginProperty> properties, CallContext context) throws PaymentPluginApiException {
+        // TODO Does Dwolla have credit?
+        return new DwollaPaymentTransactionInfoPlugin(kbPaymentId, kbTransactionId, TransactionType.CREDIT, amount, currency.toString(), Lists.newArrayList(properties));
+    }
+
+    @Override
+    public PaymentTransactionInfoPlugin refundPayment(UUID kbAccountId, UUID kbPaymentId, UUID kbTransactionId, UUID kbPaymentMethodId, BigDecimal amount, Currency currency, Iterable<PluginProperty> properties, CallContext context) throws PaymentPluginApiException {
+        // TODO send money to customer
+        // https://developers.dwolla.com/guides/send-money/
+        return null;
+    }
+
+    @Override
+    public HostedPaymentPageFormDescriptor buildFormDescriptor(UUID kbAccountId, Iterable<PluginProperty> customFields, Iterable<PluginProperty> properties, CallContext context) throws PaymentPluginApiException {
+        return null;
+    }
+
+    @Override
+    public GatewayNotification processNotification(String notification, Iterable<PluginProperty> properties, CallContext context) throws PaymentPluginApiException {
+        // TODO implement
+        return null;
+    }
+
+    public HalLink getDwollaAccount() throws PaymentPluginApiException {
+        try {
+            final CatalogResponse root = client.getRootInfo();
+            return root.getLinks().get("account");
+        } catch (ApiException e) {
+            throw new PaymentPluginApiException("There was an error loading merchant account info.", e);
+        }
+    }
+
+    private void checkValidAccessToken(final CallContext context) throws PaymentPluginApiException {
```

Comment:
A per-tenant cache would probably make sense to avoid fetching token each time from DB (we could even configure the cache in the such a way that entries expires slightly before Dwolla expires them.